In [1]:
import pandas as pd

In [2]:
df = pd.read_json("India-Food-Delivery-Time-Prediction.txt")

In [3]:
print(df.head(10))

        ID Delivery_person_ID  Delivery_person_Age  Delivery_person_Ratings  \
0  0x4607     INDORES13DEL02                  37.0                      4.9   
1  0xb379     BANGRES18DEL02                  34.0                      4.5   
2  0x5d6d     BANGRES19DEL01                  23.0                      4.4   
3  0x7a6a    COIMBRES13DEL02                  38.0                      4.7   
4  0x70a2     CHENRES12DEL01                  32.0                      4.6   
5  0x9bb4      HYDRES09DEL03                  22.0                      4.8   
6  0x95b4   RANCHIRES15DEL01                  33.0                      4.7   
7  0x9eb2      MYSRES15DEL02                  35.0                      4.6   
8  0x1102      HYDRES05DEL02                  22.0                      4.8   
9  0xcdcd      DEHRES17DEL01                  36.0                      4.2   

   Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  \
0            22.745049             75.892471          

In [4]:
print(df.shape)

(41953, 20)


In [5]:
print(df.columns)

Index(['ID', 'Delivery_person_ID', 'Delivery_person_Age',
       'Delivery_person_Ratings', 'Restaurant_latitude',
       'Restaurant_longitude', 'Delivery_location_latitude',
       'Delivery_location_longitude', 'Order_Date', 'Time_Orderd',
       'Time_Order_picked', 'Weatherconditions', 'Road_traffic_density',
       'Vehicle_condition', 'Type_of_order', 'Type_of_vehicle',
       'multiple_deliveries', 'Festival', 'City', 'Time_taken(min)'],
      dtype='str')


In [6]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 41953 entries, 0 to 41952
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           41953 non-null  str    
 1   Delivery_person_ID           41953 non-null  str    
 2   Delivery_person_Age          40234 non-null  float64
 3   Delivery_person_Ratings      40190 non-null  float64
 4   Restaurant_latitude          41953 non-null  float64
 5   Restaurant_longitude         41953 non-null  float64
 6   Delivery_location_latitude   41953 non-null  float64
 7   Delivery_location_longitude  41953 non-null  float64
 8   Order_Date                   41953 non-null  str    
 9   Time_Orderd                  40353 non-null  str    
 10  Time_Order_picked            41953 non-null  str    
 11  Weatherconditions            41953 non-null  str    
 12  Road_traffic_density         41953 non-null  str    
 13  Vehicle_condition          

suspicious column id , delivery_person_age,ratings,time_ordered

suspicious data tye date,timeordered,taken

In [7]:
print(df.isnull().sum())

ID                                0
Delivery_person_ID                0
Delivery_person_Age            1719
Delivery_person_Ratings        1763
Restaurant_latitude               0
Restaurant_longitude              0
Delivery_location_latitude        0
Delivery_location_longitude       0
Order_Date                        0
Time_Orderd                    1600
Time_Order_picked                 0
Weatherconditions                 0
Road_traffic_density              0
Vehicle_condition                 0
Type_of_order                     0
Type_of_vehicle                   0
multiple_deliveries             905
Festival                          0
City                              0
Time_taken(min)                   0
dtype: int64


In [8]:
print(df.duplicated().sum())

0


checking unique in every columns

In [9]:
print(df["Weatherconditions"].value_counts())


Weatherconditions
conditions Fog           7012
conditions Stormy        6974
conditions Cloudy        6932
conditions Sandstorms    6906
conditions Windy         6832
conditions Sunny         6728
conditions NaN            569
Name: count, dtype: int64


In [10]:
print(df["Road_traffic_density"].value_counts())

Road_traffic_density
Low        14200
Jam        13043
Medium     10084
High        4071
NaN          555
Name: count, dtype: int64


In [11]:
print(df["Type_of_order"].value_counts())

Type_of_order
Snack      10616
Meal       10524
Drinks     10445
Buffet     10368
Name: count, dtype: int64


In [12]:
print(df["Type_of_vehicle"].value_counts())

Type_of_vehicle
motorcycle           24396
scooter              14029
electric_scooter      3468
bicycle                 60
Name: count, dtype: int64


In [13]:
print(df["multiple_deliveries"].value_counts())

multiple_deliveries
1.0    25928
0.0    12952
2.0     1841
3.0      327
Name: count, dtype: int64


In [14]:
print(df["Festival"].value_counts())

Festival
No      40916
Yes       822
NaN       215
Name: count, dtype: int64


In [15]:
print(df["City"].value_counts())

City
Metropolitian     31411
Urban              9279
NaN                1114
Semi-Urban          149
Name: count, dtype: int64


In [16]:
print(df["Delivery_person_Age"].describe())

count    40234.000000
mean        29.563330
std          5.812361
min         15.000000
25%         25.000000
50%         30.000000
75%         35.000000
max         50.000000
Name: Delivery_person_Age, dtype: float64


In [17]:
print(df["Delivery_person_Ratings"].describe())


count    40190.000000
mean         4.632386
std          0.335541
min          1.000000
25%          4.500000
50%          4.700000
75%          4.900000
max          6.000000
Name: Delivery_person_Ratings, dtype: float64


In [18]:
print(df["Restaurant_latitude"].describe())

count    41953.000000
mean        18.911397
std          5.467829
min          9.957144
25%         12.986047
50%         19.065838
75%         22.751234
max         30.914057
Name: Restaurant_latitude, dtype: float64


In [19]:
print(df["Restaurant_longitude"].describe())

count    41953.000000
mean        76.923408
std          3.502910
min         72.768726
25%         73.898520
50%         76.618203
75%         78.368855
max         88.433452
Name: Restaurant_longitude, dtype: float64


In [20]:
print(df["Delivery_location_latitude"].describe())


count    41953.000000
mean        18.975064
std          5.469616
min          9.967144
25%         13.065996
50%         19.123249
75%         22.820040
max         31.054057
Name: Delivery_location_latitude, dtype: float64


In [21]:
print(df["Delivery_location_longitude"].describe())

count    41953.000000
mean        76.987074
std          3.503073
min         72.778726
25%         73.940327
50%         76.663622
75%         78.405467
max         88.563452
Name: Delivery_location_longitude, dtype: float64


In [22]:
print(df["Order_Date"].value_counts().head(10))

Order_Date
15-03-2022    1115
03-04-2022    1101
13-03-2022    1095
26-03-2022    1093
05-04-2022    1091
24-03-2022    1091
09-03-2022    1088
11-03-2022    1079
05-03-2022    1079
21-03-2022    1079
Name: count, dtype: int64


In [23]:
print(df["Time_Orderd"].value_counts().head(10))

Time_Orderd
20:32:38    8
19:56:41    8
19:21:37    7
21:16:49    7
19:02:56    7
22:59:06    7
17:37:32    7
23:31:23    7
21:42:35    7
18:11:06    7
Name: count, dtype: int64


In [24]:
print(df["Time_Order_picked"].value_counts().head(10))

Time_Order_picked
0:00:12     8
19:34:34    8
17:30:15    8
21:56:46    8
18:43:15    7
23:13:17    7
23:49:44    7
21:31:21    7
23:45:29    7
18:41:55    7
Name: count, dtype: int64


In [25]:
print(df["Vehicle_condition"].value_counts())

Vehicle_condition
1    13851
0    13815
2    13806
3      481
Name: count, dtype: int64


In [26]:
print(df["Time_taken(min)"].value_counts().head(10))

Time_taken(min)
(min) 26    1960
(min) 25    1899
(min) 27    1826
(min) 29    1795
(min) 28    1794
(min) 19    1675
(min) 15    1645
(min) 18    1604
(min) 16    1579
(min) 17    1546
Name: count, dtype: int64


In [27]:
print(df["ID"].nunique())

41953


In [28]:
df["Delivery_person_ID"].nunique()

1170

In [30]:
df["Time_taken(min)"].head()

0    (min) 24
1    (min) 33
2    (min) 26
3    (min) 21
4    (min) 30
Name: Time_taken(min), dtype: str

removing min()from column

In [31]:
df["Time_taken(min)"] = df["Time_taken(min)"].str.replace("(min) ", "")

In [32]:
df["Time_taken(min)"].head()

0    24
1    33
2    26
3    21
4    30
Name: Time_taken(min), dtype: str

In [33]:
df["Time_taken(min)"] = pd.to_numeric(df["Time_taken(min)"])

In [34]:
df["Time_taken(min)"].dtype

dtype('int64')

In [35]:
df["multiple_deliveries"].describe()

count    41048.000000
mean         0.745249
std          0.572152
min          0.000000
25%          0.000000
50%          1.000000
75%          1.000000
max          3.000000
Name: multiple_deliveries, dtype: float64

In [44]:
print(df["Vehicle_condition"].describe())

count    41953.000000
mean         1.022716
std          0.838893
min          0.000000
25%          0.000000
50%          1.000000
75%          2.000000
max          3.000000
Name: Vehicle_condition, dtype: float64


In [43]:
print(df["Time_taken(min)"].describe())

count    41953.000000
mean        26.311539
std          9.380753
min         10.000000
25%         19.000000
50%         26.000000
75%         32.000000
max         54.000000
Name: Time_taken(min), dtype: float64


In [42]:
print(df["Order_Date"].dtype)


str


In [ ]:
print(df["Order_Date"].head())


0    19-03-2022
1    25-03-2022
2    19-03-2022
3    05-04-2022
4    26-03-2022
Name: Order_Date, dtype: str

In [41]:
print(df["Order_Date"].isnull().sum())

0


In [45]:
df["Order_Date"] = pd.to_datetime(df["Order_Date"], format="%d-%m-%Y")

In [46]:
print(df["Order_Date"].head())

0   2022-03-19
1   2022-03-25
2   2022-03-19
3   2022-04-05
4   2022-03-26
Name: Order_Date, dtype: datetime64[us]


In [47]:
print(df["Time_Orderd"].dtype)

str


In [48]:
print(df["Time_Orderd"].head())

0    11:33:33
1    19:45:37
2     8:32:58
3    18:03:58
4    13:34:16
Name: Time_Orderd, dtype: str


In [49]:
print(df["Time_Orderd"].isnull().sum())

1600


In [50]:
print(df["Time_Order_picked"].dtype)

str


In [51]:
print(df["Time_Order_picked"].isnull().sum())

0


problem found till now

* Delivery_person_Age → missing (1719), suspicious (age = 15)
* Delivery_person_Ratings → missing (1763), suspicious (rating = 6)
* Time_Orderd → missing (1600)
* multiple_deliveries → missing (905)
* Vehicle_condition → suspicious (condition = 3)

handled  

* Time_taken(min) → "(min)" prefix removed, converted to int
* Order_Date → converted from str to datetime
* Duplicate rows → 0 duplicates
* Restaurant_latitude → no issue
* Restaurant_longitude → no issue
* Delivery_location_latitude → no issue
* Delivery_location_longitude → no issue
* Time_Order_picked → no missing values
* Type_of_order → no issue
* Type_of_vehicle → no issue
* Festival → no major issue identified
* City → no major issue identified
* Weatherconditions → no major issue identified
* Road_traffic_density → no major issue identified

In [52]:
print(df["Weatherconditions"].value_counts())

Weatherconditions
conditions Fog           7012
conditions Stormy        6974
conditions Cloudy        6932
conditions Sandstorms    6906
conditions Windy         6832
conditions Sunny         6728
conditions NaN            569
Name: count, dtype: int64


In [53]:
print(df["Road_traffic_density"].value_counts())

Road_traffic_density
Low        14200
Jam        13043
Medium     10084
High        4071
NaN          555
Name: count, dtype: int64
